server/src/main/java/org/elasticsearch/cluster/metadata/TemplateUpgradeService.java

```diff
@@ -151,6 +151,12 @@ void updateTemplates(Map<String, BytesReference> changes, Set<String> deletions)
             client.admin().indices().putTemplate(request, new ActionListener<PutIndexTemplateResponse>() {
                 @Override
                 public void onResponse(PutIndexTemplateResponse response) {
+                    // check successful template change
+                    final Optional<Tuple<Map<String, BytesReference>, Set<String>>> afterChanges =
+                            calculateTemplateChanges(clusterService.state().getMetaData().getTemplates());
+                    if (afterChanges.isPresent() && afterChanges.get().v1().containsKey(request.name())) {
+                        throw new IllegalStateException("Successful template [" + request.name() + "] upgrade is not successful");
```

Comment:
I think this error message is going to be very confusing if anyone runs into it. "Successful" is not "successful" ?

Maybe
> Template [{}] was updated successfully, but is still being reported as out of date.

However, I don't think throwing an exception in `onResponse` makes sense (I know it's based on feedback from @jaymode) What catches this exception? How does it get handled?

Reply:
Thanks Tim, good point about throwing exception in `onResponse` handlers.